core/whistle_number_manager-1.0.0/src/wh_port_request.erl

```diff
@@ -155,60 +158,71 @@ charge_for_port(_JObj, AccountId) ->
     Transaction = wh_transaction:debit(AccountId, Cost),
     wh_services:commit_transactions(Services, [Transaction]).
 
--spec completed_port(wh_json:object()) ->
+-spec completed_port(ne_binary(), wh_json:object()) ->
                             transition_response().
-completed_port(PortReq) ->
+completed_port(PriorState, PortReq) ->
     case charge_for_port(PortReq) of
         'ok' ->
             lager:debug("successfully charged for port, transitioning numbers to active"),
-            transition_numbers(PortReq);
+            transition_numbers(PriorState, PortReq);
         'error' ->
-            throw({'error', 'failed_to_charge'})
+            {'error', 'failed_to_charge'}
     end.
 
--spec transition_numbers(wh_json:object()) ->
+-spec transition_numbers(ne_binary(), wh_json:object()) ->
                                 transition_response().
-transition_numbers(PortReq) ->
-    Numbers = wh_json:get_keys(<<"numbers">>, PortReq),
-    PortOps = [enable_number(N) || N <- Numbers],
-    case lists:all(fun(X) -> wh_util:is_true(X) end, PortOps) of
-        'true' ->
-            lager:debug("all numbers ported, removing from port request"),
-            ClearedPortRequest = clear_numbers_from_port(PortReq),
-            {'ok', ClearedPortRequest};
-        'false' ->
-            lager:debug("failed to transition numbers: ~p", [PortOps]),
-            {'error', PortReq}
-    end.
+transition_numbers(PriorState, JObj) ->
+    finalize_number_transition(
+      PriorState
+      ,enable_numbers(
+         wh_json:get_keys(<<"numbers">>, JObj)
+         ,JObj
+         ,[]
+        )
+     ).
 
--spec clear_numbers_from_port(wh_json:object()) -> wh_json:object().
-clear_numbers_from_port(PortReq) ->
-    case couch_mgr:save_doc(?KZ_PORT_REQUESTS_DB
-                            ,wh_json:set_value(<<"numbers">>, wh_json:new(), PortReq)
-                           )
-    of
-        {'ok', PortReq1} -> lager:debug("port numbers cleared"), PortReq1;
-        {'error', 'conflict'} ->
-            lager:debug("port request doc was updated before we could re-save"),
-            PortReq;
-        {'error', _E} ->
-            lager:debug("failed to clear numbers: ~p", [_E]),
-            PortReq
+-spec finalize_number_transition(ne_binary(), {wh_json:object(), wh_proplist()}) -> transition_response().
+finalize_number_transition(_, {JObj, []}) ->
+    {'ok', Updated} = couch_mgr:ensure_saved(?KZ_PORT_REQUESTS_DB, JObj),
+    {'ok', Updated};
```

Comment:
You can pattern match and leave it as the last expression; no need to re-create the tuple again to return it.